src/smt/smt_engine_state.h

```diff
@@ -0,0 +1,240 @@
+   * is issued to an SmtEngine, it may use a satsisfiability call (if desired)
+   * to solve the abduction query. This method is called *in addition* to
+   * the above calls to notifyCheckSat / notifyCheckSatResult in this case.
+   */
+  void notifyGetAbduct(bool success);
+  /**
+   * Intialize the context, which makes a single push to maintain a global
+   * context around everything.
+   */
+  void initialize();
+  /**
+   * Prepare for a shutdown of the SmtEngine, which does pending pops and
+   * pops the user context to zero.
+   */
+  void shutdown();
+  /**
+   * Cleanup, which pops all contexts to level zero.
+   */
+  void cleanup();
+  /**
+   * Do all pending pops, which ensures that the context levels are up-to-date.
+   * This method should be called by the SmtEngine before using any of its
+   * members that rely on the context (e.g. PropEngine or TheoryEngine).
+   */
+  void doPendingPops();
+  /**
+   * Set that the file name of the current instance is the given string. This
+   * is used for various purposes (e.g. get-info, SZS status).
+   */
+  void setFilename(std::string filename);
+
+  //---------------------------- context management
+  /**
+   * Called when the user of SmtEngine issues a push. This corresponds to
+   * the SMT-LIB command push.
+   */
+  void userPush();
+  /**
+   * Called when the user of SmtEngine issues a pop. This corresponds to
+   * the SMT-LIB command pop.
+   */
+  void userPop();
+  /** Get a pointer to the UserContext owned by this SmtEngine. */
+  context::UserContext* getUserContext();
+  /** Get a pointer to the Context owned by this SmtEngine. */
+  context::Context* getContext();
+  //---------------------------- end context management
+
+  //---------------------------- queries
+  /**
+   * Return true if the SmtEngine is fully initialized (post-construction).
+   * This post-construction initialization is automatically triggered by the
+   * use of the SmtEngine; e.g. when the first formula is asserted, a call
+   * to simplify() is issued, a scope is pushed, etc.
+   */
+  bool isFullyInited() const;
+  /**
+   * Return true if the SmtEngine is fully initialized and there are no
+   * pending pops.
+   */
+  bool isFullyReady() const;
+  /**
+   * Return true if a notifyCheckSat call has been made, e.g. a query has been
+   * issued to the SmtEngine.
+   */
+  bool isQueryMade() const;
+  /** Return the user context level.  */
+  size_t getNumUserLevels() const;
+  /** Get the status of the last check-sat */
+  Result getStatus() const;
+  /** Get the SMT mode we are in */
+  SmtMode getMode() const;
+  /** return the input name (if any) */
+  std::string getFilename() const;
+  //---------------------------- end queries
+
+ private:
+  /** Pushes the user and SAT contexts */
+  void push();
+  /** Pops the user and SAT contexts */
+  void pop();
+  /** Pops the user and SAT contexts to the given level */
+  void popto(int toLevel);
```

Comment:
How about:

```suggestion
  void popTo(size_t toLevel);
```

(to mirror the return type of `getNumUserLevels()`)

Reply:
This is a wrapper method to Context::popto which takes `int` unfortunately, I believe we should keep it like this until that method is change.